agents/cache/cache.go

```diff
@@ -0,0 +1,179 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"container/list"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// ErrWrongType is returned by different functions when the type of the passed argument cannot be used to create or manipulate Cache.
+var ErrWrongType = errors.New("wrong argument type")
+
+// Cache provides cached access to performance statistics tables.
+// It retains data longer than those tables.
+// Intended to store various subtypes of map.
+type Cache struct {
+	typ       reflect.Type
+	retain    time.Duration
+	sizeLimit uint
+	l         *logrus.Entry
+
+	rw        sync.RWMutex
+	items     map[interface{}]*list.Element
+	itemsList *list.List
+	updatedN  uint
+	addedN    uint
+	removedN  uint
+}
+
+// cacheItem is an element stored in Cache
+type cacheItem struct {
+	key   interface{}
```

Comment:
> do we really need this field?
We need it for backward relation from List to map.

> what types can be used as key? I though it's always string in our cases.
It would be much easier case and we would use one of existing open source lib. But we also use int64 keys.